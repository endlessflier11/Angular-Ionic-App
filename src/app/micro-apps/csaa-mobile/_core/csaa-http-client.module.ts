import { NgModule } from '@angular/core';
import { AngularHttpAdapter } from './services/http/adapters/angular-http.adapter';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { IonicHttpAdapter } from './services/http/adapters/ionic-http.adapter';
import { HTTP } from '@ionic-native/http/ngx';
import { UNI_HTTP_ADAPTER } from './services/http/uni-http.model';

@NgModule({
  imports: [HttpClientModule],
  providers: [
    HTTP,
    { provide: UNI_HTTP_ADAPTER, multi: true, useClass: AngularHttpAdapter, deps: [HttpClient] },
    {
      provide: UNI_HTTP_ADAPTER,
      multi: true,
      useClass: IonicHttpAdapter,
      deps: [HTTP],
    },
  ],
  exports: [HttpClientModule],
})
export class CsaaHttpClientModule {}
