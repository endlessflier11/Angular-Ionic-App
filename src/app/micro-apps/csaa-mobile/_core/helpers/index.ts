export * from './router.helpers';
export * from './misc.helpers';
export * from './async.helper';
export * from './error.helper';
export * from './string.helper';
