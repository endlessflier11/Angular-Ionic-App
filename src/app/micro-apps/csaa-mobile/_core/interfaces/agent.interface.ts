export interface Agent {
  name: string;
  phone: string;
  email: string;
  isService: boolean;
}
