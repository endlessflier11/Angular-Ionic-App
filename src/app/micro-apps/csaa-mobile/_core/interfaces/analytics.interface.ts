export enum Category {
  landing,
  application,
  claims,
  coverages,
  documents,
  global,
  payments,
  registration,
  error,
  debug,
}

export enum EventName {
  INSURANCE_TAB_TAPPED = 'Insurance Tab Tapped',
  LANDING_PAGE_MAKE_A_PAYMENT_SELECTED = 'Landing Page Make a Payment Selected',
  LANDING_PAGE_MANAGE_MY_POLICY_SELECTED = 'Landing Page Manage My Policy Selected',
  LANDING_PAGE_CALL_FOR_QUOTE_SELECTED = 'Landing Page Call For Quote Selected',
  LANDING_PAGE_START_A_CLAIM_SELECTED = 'Landing Page Start A Claim Selected',
  LANDING_PAGE_WEB_QUOTE_SELECTED = 'Landing Page Web Quote Selected',
  I_HAD_AN_ACCIDENT_SELECTED = 'I Had An Accident Selected',
  GET_A_QUOTE_SELECTED = 'Get a Quote Selected',
  VIEW_DECLARATIONS = 'View Declarations',
  CAMERA_ACCESSED = 'Camera Accessed',
  CALL_911 = 'Call 911',
  CONTACT_CLAIMS = 'Contact Claims',
  CONTACT_INITIATED = 'Contact Initiated',
  OPEN_CLAIMS_VIEWED = 'Open Claims Viewed',
  CARD_EXPANDED = 'Card Expanded',
  WHAT_DO_I_DO_SELECTED = 'What Do I Do Selected',
  I_HAD_AN_INCIDENT_SELECTED = 'I had an incident selected',
  CLAIM_SELECTED = 'Claim Selected',
  COVERAGES_VIEWED = 'Coverages Viewed',
  COVERAGE_SELECTED = 'Coverage Selected',
  AGENT_CONTACTED = 'Agent contacted',
  VEHICLE_COVERAGE_SELECTED = 'Vehicle Coverage Selected',
  DRIVER_COVERAGE_SELECTED = 'Driver Coverage Selected',
  GENERAL_COVERAGE_SELECTED = 'General Coverage Selected',
  PROOF_OF_INSURANCE_VIEWED = 'Proof of Insurance Viewed',
  AUTO_ID_CARDS_ACCESSED = 'Auto ID Cards Accessed',
  VEHICLE_ID_CARD_EXPANDED = 'Vehicle ID Card Expanded',
  CALL_SUPPORT = 'Call support',
  HOME_ACCESSED = 'Home Accessed',
  TRANSACTION_HISTORY_PAGE_SELECTED = 'Transaction History Page Selected',
  PAYMENTS_ACCESSED = 'Payments Accessed',
  PAYMENT_METHOD_ADDED = 'Payment Method Added',
  PAYMENT_METHOD_UPDATED = 'Payment Method Updated',
  PAYMENT_AMOUNT_CANCELLED = 'Payment Amount Cancelled',
  TERMS_AND_CONDITIONS_CLICKED = 'Terms and Conditions Clicked',
  TERMS_AND_CONDITIONS_ANSWERED = 'Terms and Conditions Answered',
  PAYMENT_CONFIRMED = 'Payment Confirmed',
  PAYMENT_SUCCESSFUL_DETAILS = 'Payment Successful Details',
  PAYMENT_AMOUNT_CONFIRMED = 'Payment Amount Confirmed',
  PAYMENT_PAST_DUE_NOTICE = 'Payment Past Due Notice',
  TRANSACTION_HISTORY_ACCESSED = 'Transaction History Accessed',
  PAYMENT_FAILED = 'Payment Failed',
  MAKE_A_PAYMENT_SELECTED = 'Make a Payment Selected',
  PAY_ALL_SELECTED = 'Pay All Selected',
  SIGN_UP_SELECTED = 'Sign up selected',
  AUTO_POLICY_SELECTED = 'Auto policy selected',
  HOME_POLICY_SELECTED = 'Home policy selected',
  PROPERTY_POLICY_SELECTED = 'Property policy selected',
  VERIFICATION_INITIATED = 'Verification initiated',
  VERIFICATION_REQUESTED = 'Verification requested',
  VERIFICATION_SENT = 'Verification sent',
  VERIFICATION_RECEIVED = 'Verification received',
  VERIFICATION_FAILED = 'Verification failed',
  VERIFICATION_SUCCEEDED = 'Verification succeeded',
  CALL_SERVICE_ERROR_MESSAGE = 'Call Service Error Message',
  NEW_CODE_INITIATED = 'New code initiated',
  PIN_CREATED = 'Pin created',
  SIGN_OUT_SELECTED = 'Sign out selected',
  USER_SIGNED_OUT = 'User signed out',
  SIGN_OUT_CANCELED = 'Sign out canceled',
  FILE_A_CLAIM_SELECTED = 'File a Claim Selected',
  MANAGE_AUTOPAY_VIEWED = 'Manage AutoPay Viewed',
  AUTOPAY_SET = 'AutoPay Set',
  AUTOPAY_ENROLLMENT_STOPPED = 'AutoPay Enrollment Stopped',
  MYPOLICY_LINK_ACCESSED = 'MyPolicy Link Accessed',
  DOCUMENTS_ACCESSED = 'Documents Accessed',
  DOCUMENT_SELECTED = 'Document Selected',
  DOCUMENT_VIEWED = 'Document Viewed',
  DOCUMENT_SAVED = 'Document Saved',
  API_ERROR = 'Api Error',
  DEBUG = 'Debug',
  ERROR_NOTIFICATION = 'Error Notification',
  WARNING_DISPLAYED = 'Warning Displayed',
  WARNING_RESPONSE_RECEIVED = 'Warning Response Received',
  PAPERLESS_TERMS_AND_CONDITIONS_REVIEWED = 'Paperless Terms and Conditions Reviewed',
  PAPERLESS_TERMS_AND_CONDITIONS_ACCEPTED = 'Paperless Terms and Conditions Accepted',
  PAPERLESS_SCREEN_VIEWED = 'Paperless Screen Viewed',
  PAPERLESS_ENROLLMENT_SELECTED = 'Paperless Enrollment Selected',
  EDIT_PAPERLESS_SELECTED = 'Edit Paperless Selected',
  PAPERLESS_CARD_EXPANDED = 'Paperless Card Expanded',
  AUTOPAY_INSTALLMENT_TABLE_EXPANDED = 'AutoPay Installment Table Expanded',
  SAVE_WITH_AUTOPAY_SELECTED = 'Save With AutoPay Selected',
  REVIEW_TERMS_AND_CONDITIONS_SELECTED = 'Review Terms & Conditions Selected',
  ADD_TO_APPLE_WALLET_SELECTED = 'Add to Apple Wallet Selected',
}

export enum EventType {
  OPTION_SELECTED = 'Option Selected',
  LINK_ACCESSED = 'Link Accessed',
  FILE_DOWNLOADED = 'File Downloaded',
  MESSAGED = 'Messaged',
  AUTOMATED_SYSTEM_PROCESS = 'Automated System Process',
  USER_INFORMATION_ENTERED = 'User Information Entered',
  UX_MODIFIED = 'UX Modified',
}

export interface Event {
  category: Category;
  event: EventName;
  // eslint-disable-next-line @typescript-eslint/ban-types
  properties?: object;
}

export enum Screens {
  'Sign Up',
  'Verification',
  'Passcode Creation',
  'Passcode Reset',
  'Passcode Re-Entry',
  'Verification Code',
  // 'PIN Creation', // No such screen - passcode is the right term
  // 'PIN Re-Entry',
  'Bio-Auth Setup',
  'Login',
  'Security',
  'Account Information',
  'Payments',
  'Select an Amount',
  'Add Payment Method',
  'Payment Method',
  'Payment Method Detail',
  'Settings',
  'Policy Page',
  'Various',
  'Re-Register',
  'Home',
  'Payment History',
  'Select Autopay',
  'Autopay Method',
  'Autopay Method Detail',
  'Claims',
  'Policy Document',
  'Policy Summary',
  'Legal',
  'Edit Email',
  'Edit Phone',
  'Email Updated',
  'Phone Updated',
}
