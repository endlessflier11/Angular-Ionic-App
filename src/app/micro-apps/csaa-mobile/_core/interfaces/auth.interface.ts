export enum WebDeeplinkLocation {
  DASHBOARD = 'dashboard',
  TRANSACTIONS = 'transactions',
  CLAIMS = 'claims',
  ENDORSMENT = 'endorsement',
  ACCOUNT_SETTINGS = 'accountSettings',
  EDIT_POLICY_PAPERLESS_PREFERENCES = 'accountSettings',
  // EDIT_POLICY_PAPERLESS_PREFERENCES = 'accountSettings/policy/{policyNumber}/preferences',
  // EDIT_POLICY = 'accountSettings/policy/{policyNumber}',
  GET_WALLET_PASS = 'v1/passes/{walletPassId}/{policyNumber}',
}

export enum PaperlessEnrollmentStatus {
  pending = 'pending',
  accepted = 'accepted',
}

export interface User {
  hash: string;
}

export interface CustomerSearchResponse {
  mdmId: string;
  email: string;
  fullName: string;
  firstName: string;
  lastName: string;
  customerAddress: {
    addressType: string;
    address1: string;
    address2?: string;
    zipCode: string;
    city: string;
    state: string;
    isPoBox?: boolean;
  };
  insuranceAddress: {
    address2: string;
    additional: string;
    addressType: string;
    address1: string;
    zipCode: string;
    city: string;
    state: string;
  };
  policies: CustomerSearchPolicyResponse[];
  hash: string;
  registrations?: Registration[];
  isPaperlessEnrollmentPending?: boolean;
}

export interface RoleResponse {
  sourceId: string;
  roleType: string;
  roleStatus: string;
}

export interface CustomerSearchPolicyResponse {
  policyNumber: string;
  policyStatus: string;
  policyType: string;
  clubCode: string;
  riskState: string;
  prodTypeCode: string;
  termExpirationDate: string;
  roleList: RoleResponse[];
  insuranceAddress: {
    address1: string;
    address2: string;
    additional: string;
    addressType: string;
    zipCode: string;
    city: string;
    state: string;
    siteDomain: string;
    phone: string;
    streetAddressLine: string;
    stateFullName: string;
  };
}

export interface Registration {
  registrationId?: string;
  registrationSource?: string;
  startDate?: string;
}

export interface MwgSsoAuthResponse {
  accessToken: string;
  idToken: string;
  state: string;
  expiresIn: number;
}

export interface AAAOAuthAuthResponse {
  code: string;
  clubcode: string;
  state: string;
}

export interface WebDeeplinkInitializationPayload {
  pageURI: WebDeeplinkLocation | string;
  accessToken: string;
  idToken: string;
}

export interface WebDeeplinkInitializationResponse {
  tempToken: string;
  expiresIn: number;
}
