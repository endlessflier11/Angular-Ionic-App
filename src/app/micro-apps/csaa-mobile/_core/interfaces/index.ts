export * from './auth.interface';
export * from './router';
export * from './http.interface';
export * from './vehicle.interface';
export * from './ionic-lifecycle.interface';
export * from './analytics.interface';
export * from './policy.interface';
export * from './payment.interface';
export * from './config.interface';
export * from './ngxs-store.interface';
