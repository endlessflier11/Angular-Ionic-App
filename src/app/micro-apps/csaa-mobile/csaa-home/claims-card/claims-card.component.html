<csaa-skeleton-card *ngIf='loading'></csaa-skeleton-card>
<ion-card *ngIf='!loading' class='csaa-card'>
  <ion-card-header>
    <ion-item>
      <ion-card-title> Claims</ion-card-title>
      <ion-button slot='end' fill='clear' (click)='openWhatDoIDo()'>What do I do?</ion-button>
    </ion-item>
  </ion-card-header>
  <ion-card-content class='csaa-item-list'>
    <ion-list *ngIf='sortedClaims?.length > 0'>
      <ng-container *ngFor='let claim of sortedClaims'>
        <ion-item detail (click)='openDetails(claim)' *ngIf='claim'>
          <ion-label>
            <!-- Line 1 -->
            <h3>
              <ng-container *ngIf='claim.type === PolicyType.PUP'> Liability Claim</ng-container>
              <ng-container *ngIf='claim.type !== PolicyType.PUP'>
                {{ claim.cause ? claim.cause : 'Claim #' + claim.number }}
              </ng-container>
            </h3>
            <!-- Line 2 -->
            <ng-container *ngIf='claim.type === PolicyType.PUP || claim.cause'>
              <span class='claim-number'>Claim #{{ claim.number }}</span>
              <br />
            </ng-container>

            <ng-container *ngIf='claim.status'>
              <span class='status'>Status: {{ claim.status }}</span>
              <br />
            </ng-container>

            <!-- Line 3 -->
            <span class='subtitle'>
            <ng-container *ngIf='claim.vehicle && claim.type === PolicyType.Auto'>
              {{ claim.vehicle.name }}
            </ng-container>
            <ng-container *ngIf='claim.type === PolicyType.Home'>
              {{ claim.address }}
            </ng-container>
            <ng-container *ngIf='claim.type === PolicyType.PUP'>
              Personal Umbrella Policy
            </ng-container>
          </span>
          </ion-label>

          <csaa-theme-icon slot='end' [type]='claim.type' [size]='40'></csaa-theme-icon>
        </ion-item>
      </ng-container>
    </ion-list>
    <ion-button expand='full' (click)='openIHadAnAccident()'>I Had an Accident</ion-button>
  </ion-card-content>
</ion-card>
