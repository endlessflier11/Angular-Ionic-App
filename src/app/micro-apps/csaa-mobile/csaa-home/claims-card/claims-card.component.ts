import { Component, Input, OnChanges, SimpleChanges } from '@angular/core';
import { AlertController } from '@ionic/angular';
import { CLAIMS_NUMBER } from '../../constants';
import { Category, EventName, EventType } from '../../_core/interfaces/analytics.interface';
import { Claim } from '../../_core/interfaces/claim.interface';
import { Policy, PolicyType } from '../../_core/interfaces/policy.interface';
import { AnalyticsService } from '../../_core/services/analytics.service';
import { CallService } from '../../_core/services/call.service';
import { RouterService } from '../../_core/services/router/router.service';

@Component({
  selector: 'csaa-claims-card',
  templateUrl: './claims-card.component.html',
  styleUrls: ['./claims-card.component.scss'],
})
export class ClaimsCardComponent implements OnChanges {
  @Input() claims: Claim[] = [];
  @Input() policies: Policy[] = [];
  @Input() loading = false;

  // eslint-disable-next-line @typescript-eslint/naming-convention
  PolicyType = PolicyType;
  sortedClaims: Claim[] = [];

  constructor(
    private analyticsService: AnalyticsService,
    private alertCtrl: AlertController,
    private callService: CallService,
    private routerService: RouterService
  ) {}

  ngOnChanges(changes: SimpleChanges) {
    if (changes.claims && changes.claims.currentValue && changes.claims.currentValue.length) {
      this.sortedClaims = this.sortClaimsByOpenedDate(this.claims.slice());
    }
  }

  private sortClaimsByOpenedDate(claims: Claim[]) {
    const getOpenedDateFromWorkflow = (claim) => {
      if (!claim?.workflow) {
        return 0;
      }
      const openedClaim = claim?.workflow?.find((event) => event.eventType === 'Opened');
      return openedClaim?.date ? new Date(openedClaim?.date) : 0;
    };

    return claims?.slice().sort((first, second) => {
      const firstOpenedDate: any = getOpenedDateFromWorkflow(first);
      const secondOpenedDate: any = getOpenedDateFromWorkflow(second);
      return secondOpenedDate - firstOpenedDate;
    });
  }

  trackAgentContact(selection: 'Contact Claims') {
    this.analyticsService.trackEvent(EventName.CONTACT_INITIATED, Category.claims, {
      event_type: EventType.OPTION_SELECTED,
      method: 'phone',
      selection,
    });
  }

  async openDetails(claim) {
    if (claim.type === PolicyType.PUP) {
      const alert = await this.alertCtrl.create({
        header: 'Call Claims Representative',
        message: claim.adjuster.phone,
        buttons: [
          {
            text: 'Cancel',
            role: 'cancel',
            handler: () => {
              alert.dismiss(false);
              return false;
            },
          },
          {
            text: 'Call',
            handler: () => {
              this.trackAgentContact('Contact Claims');
              alert.dismiss(true);
              return false;
            },
          },
        ],
      });
      await alert.present();
      alert.onDidDismiss().then(({ role }) => {
        if (role === 'cancel') {
          return;
        }
        this.callService.call(claim.adjuster.phone || CLAIMS_NUMBER);
      });
    } else {
      const policies = this.policies.filter((p) => p.number === claim.policyNumber);
      this.analyticsService.trackEvent(EventName.CLAIM_SELECTED, Category.claims, {
        claim_number: claim.number,
        ...AnalyticsService.mapPolicy(...policies),
        event_type: EventType.OPTION_SELECTED,
        selection: 'Claims',
      });
      this.analyticsService.trackEvent(EventName.OPEN_CLAIMS_VIEWED, Category.claims, {
        claim_number: claim.number,
        ...AnalyticsService.mapPolicy(...policies),
        event_type: EventType.LINK_ACCESSED,
        link: 'Claims',
      });
      await this.routerService.navigateForward('csaa.claims.detail', {
        claimNumber: claim.number,
      });
    }
  }

  openWhatDoIDo() {
    this.analyticsService.trackEvent(EventName.WHAT_DO_I_DO_SELECTED, Category.claims, {
      event_type: EventType.OPTION_SELECTED,
      selection: 'What To Do',
    });
    this.whatDoIDo();
  }

  openIHadAnAccident() {
    this.analyticsService.trackEvent(EventName.I_HAD_AN_ACCIDENT_SELECTED, Category.claims, {
      event_type: EventType.OPTION_SELECTED,
      selection: 'I Had An Accident',
    });
    const hasAutoPolicies = !!this.policies.filter((p) => p.type === PolicyType.Auto).length;
    if (hasAutoPolicies) {
      this.routerService.navigateForward('csaa.claims.what-do-do', {
        policyType: PolicyType.Auto.toString(),
      });
    } else {
      this.whatDoIDo();
    }
  }

  private whatDoIDo() {
    const hasAutoPolicies = !!this.policies.filter((p) => p.type === PolicyType.Auto).length;
    const hasHomePolicies = !!this.policies.filter((p) => p.type === PolicyType.Home).length;
    if (hasAutoPolicies && !hasHomePolicies) {
      this.routerService.navigateForward('csaa.claims.what-do-do', {
        policyType: PolicyType.Auto.toString(),
      });
    } else if (hasHomePolicies && !hasAutoPolicies) {
      this.routerService.navigateForward('csaa.claims.what-do-do', {
        policyType: PolicyType.Home.toString(),
      });
    } else {
      this.routerService.navigateForward('csaa.claims.pre-loss');
    }
  }
}
