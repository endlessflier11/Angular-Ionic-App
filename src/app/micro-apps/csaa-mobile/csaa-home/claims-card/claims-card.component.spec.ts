import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { IonicModule } from '@ionic/angular';
import { IonicStorageModule } from '@ionic/storage';
import { generatePolicy, PageTestingModule } from '../../../../../testing';

import { Category, EventName } from '../../_core/interfaces';
import { Claim } from '../../_core/interfaces/claim.interface';
import { Policy } from '../../_core/interfaces/policy.interface';
import { AnalyticsService } from '../../_core/services/analytics.service';
import { CallService } from '../../_core/services/call.service';

import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';

import { ClaimsCardComponent } from './claims-card.component';
import { Component, ViewChild } from '@angular/core';

@Component({
  // eslint-disable-next-line @angular-eslint/component-selector
  selector: `test-host-component`,
  template: `<div>
    <csaa-claims-card [claims]="claims" [policies]="policies" [loading]="loading">
    </csaa-claims-card>
  </div>`,
})
export class ClaimsCardHostComponent {
  @ViewChild(ClaimsCardComponent)
  public claimsCardComponent: any;
  public claims: Claim[] = [];
  public policies: Policy[] = [];
  public loading = false;
}

describe('ClaimsCardComponent', () => {
  let claim: Claim;
  let policy: Policy;
  let component: ClaimsCardComponent;
  let fixture: ComponentFixture<ClaimsCardComponent>;
  let analyticsService: AnalyticsService;

  beforeEach(async(() => {
    try {
      TestBed.configureTestingModule({
        declarations: [ClaimsCardComponent],
        providers: [{ provide: CallService, useValue: { call: jest.fn() } }],
        imports: [
          PageTestingModule.withConfig({
            providesAlert: true,
            providesRouter: true,
            providesAnalytics: true,
          }),
          IonicModule,
          IonicStorageModule.forRoot(),
          UiKitsModule,
        ],
      }).compileComponents();

      policy = generatePolicy({
        vehicles: [
          {
            id: 'WAUKJAFM8AA061319',
            name: 'Lamborghini Gallardo',
            coverages: [],
            riskFactors: { waivedLiability: true },
          },
          {
            id: 'WMWZC3C51EWP29276',
            name: 'Aston Martin DB9',
            coverages: [],
            riskFactors: { waivedLiability: false },
          },
        ],
      });

      claim = {
        number: '12345',
        status: 'active',
        cause: 'test',
        type: policy.type,
        policyNumber: policy.number,
        vehicle: policy.vehicles[0],
        workflow: [],
        address: policy.address,
      };

      analyticsService = TestBed.inject(AnalyticsService);
      fixture = TestBed.createComponent(ClaimsCardComponent);
      component = fixture.componentInstance;
      component.policies = [policy];
      fixture.detectChanges();
    } catch (error) {
      console.error(error);
      fail(error.message);
    }
  }));

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should track claims card', () => {
    component.openIHadAnAccident();
    expect(analyticsService.trackEvent).toHaveBeenCalledWith(
      EventName.I_HAD_AN_ACCIDENT_SELECTED,
      Category.claims,
      {
        event_type: 'Option Selected',
        selection: 'I Had An Accident',
      }
    );
  });
  it('should track the claims selected', async () => {
    await component.openDetails(claim);
    expect(analyticsService.trackEvent).toHaveBeenCalledWith(
      EventName.CLAIM_SELECTED,
      Category.claims,
      {
        claim_number: claim.number,
        policies: [
          {
            policy_number: '1',
            policy_state: undefined,
            policy_type: 'Auto',
          },
        ],
        event_type: 'Option Selected',
        selection: 'Claims',
      }
    );
  });
  it('should track the card details', async () => {
    await component.openDetails(claim);
    expect(analyticsService.trackEvent).toHaveBeenCalledWith(
      EventName.OPEN_CLAIMS_VIEWED,
      Category.claims,
      {
        claim_number: claim.number,
        link: 'Claims',
        policies: [
          {
            policy_number: '1',
            policy_state: undefined,
            policy_type: 'Auto',
          },
        ],
        event_type: 'Link Accessed',
      }
    );
  });

  it('should track What Do I Do ', () => {
    component.openWhatDoIDo();
    expect(analyticsService.trackEvent).toHaveBeenCalledWith(
      EventName.WHAT_DO_I_DO_SELECTED,
      Category.claims,
      {
        event_type: 'Option Selected',
        selection: 'What To Do',
      }
    );
  });
  it('should track the track agent contact', () => {
    component.trackAgentContact('Contact Claims');
    expect(analyticsService.trackEvent).toHaveBeenCalledWith(
      EventName.CONTACT_INITIATED,
      Category.claims,
      {
        event_type: 'Option Selected',
        method: 'phone',
        selection: 'Contact Claims',
      }
    );
  });
});

describe('ClaimsCardComponent from Host', () => {
  let claim: Claim;
  let policy: Policy;
  let componentHost: ClaimsCardHostComponent;

  let fixture: ComponentFixture<ClaimsCardHostComponent>;

  beforeEach(async(() => {
    try {
      TestBed.configureTestingModule({
        declarations: [ClaimsCardHostComponent, ClaimsCardComponent],
        providers: [{ provide: CallService, useValue: { call: jest.fn() } }],
        imports: [
          PageTestingModule.withConfig({
            providesAnalytics: true,
            providesRouter: true,
            providesConfig: true,
            providesAlert: true,
          }),
          IonicModule,
          IonicStorageModule.forRoot(),
          UiKitsModule,
        ],
      }).compileComponents();

      policy = generatePolicy({
        vehicles: [
          {
            id: 'WAUKJAFM8AA061319',
            name: 'Lamborghini Gallardo',
            coverages: [],
            riskFactors: { waivedLiability: true },
          },
          {
            id: 'WMWZC3C51EWP29276',
            name: 'Aston Martin DB9',
            coverages: [],
            riskFactors: { waivedLiability: false },
          },
        ],
      });

      claim = {
        number: '12345',
        status: 'active',
        cause: 'test',
        type: policy.type,
        policyNumber: policy.number,
        vehicle: policy.vehicles[0],
        workflow: [],
        address: policy.address,
      };

      fixture = TestBed.createComponent(ClaimsCardHostComponent);
      componentHost = fixture.componentInstance;
      componentHost.policies = [policy];
      componentHost.claims = [claim];
      fixture.detectChanges();
    } catch (error) {
      console.error(error);
      fail(error.message);
    }
  }));

  it('should create host component', () => {
    expect(componentHost).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should not fail ngOnChanges if got null claims', () => {
    try {
      componentHost.claims = [claim, null, undefined, claim];
      fixture.detectChanges();
      expect(componentHost).toBeTruthy();
      expect(fixture).toMatchSnapshot();
    } catch (e) {
      console.log(e);
      fail(e.message);
    }
  });
});
