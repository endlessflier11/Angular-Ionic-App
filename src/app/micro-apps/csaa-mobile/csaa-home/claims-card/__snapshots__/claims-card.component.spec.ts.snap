// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaimsCardComponent from Host should create host component 1`] = `
<test-host-component
  __ngContext__={[Function LRootView]}
  claims={[Function Array]}
  claimsCardComponent={[Function ClaimsCardComponent]}
  loading="false"
  policies={[Function Array]}
>
  <div>
    <csaa-claims-card>
      <ion-card
        class="csaa-card"
      >
        <ion-card-header>
          <ion-item>
            <ion-card-title>
               Claims
            </ion-card-title>
            <ion-button
              fill="clear"
              slot="end"
            >
              What do I do?
            </ion-button>
          </ion-item>
        </ion-card-header>
        <ion-card-content
          class="csaa-item-list"
        >
          <ion-list>
            <ion-item
              detail=""
            >
              <ion-label>
                <h3>
                   test 
                </h3>
                <span
                  class="claim-number"
                >
                  Claim #12345
                </span>
                <br />
                <span
                  class="status"
                >
                  Status: active
                </span>
                <br />
                <span
                  class="subtitle"
                >
                   Lamborghini Gallardo 
                </span>
              </ion-label>
              <csaa-theme-icon
                slot="end"
              >
                <img
                  src="/assets/csaa-mobile/vectors/icn-Auto.svg"
                  style="width: 40px; height: 40px;"
                />
              </csaa-theme-icon>
            </ion-item>
          </ion-list>
          <ion-button
            expand="full"
          >
            I Had an Accident
          </ion-button>
        </ion-card-content>
      </ion-card>
    </csaa-claims-card>
  </div>
</test-host-component>
`;

exports[`ClaimsCardComponent from Host should not fail ngOnChanges if got null claims 1`] = `
<test-host-component
  __ngContext__={[Function LRootView]}
  claims={[Function Array]}
  claimsCardComponent={[Function ClaimsCardComponent]}
  loading="false"
  policies={[Function Array]}
>
  <div>
    <csaa-claims-card>
      <ion-card
        class="csaa-card"
      >
        <ion-card-header>
          <ion-item>
            <ion-card-title>
               Claims
            </ion-card-title>
            <ion-button
              fill="clear"
              slot="end"
            >
              What do I do?
            </ion-button>
          </ion-item>
        </ion-card-header>
        <ion-card-content
          class="csaa-item-list"
        >
          <ion-list>
            <ion-item
              detail=""
            >
              <ion-label>
                <h3>
                   test 
                </h3>
                <span
                  class="claim-number"
                >
                  Claim #12345
                </span>
                <br />
                <span
                  class="status"
                >
                  Status: active
                </span>
                <br />
                <span
                  class="subtitle"
                >
                   Lamborghini Gallardo 
                </span>
              </ion-label>
              <csaa-theme-icon
                slot="end"
              >
                <img
                  src="/assets/csaa-mobile/vectors/icn-Auto.svg"
                  style="width: 40px; height: 40px;"
                />
              </csaa-theme-icon>
            </ion-item>
            <ion-item
              detail=""
            >
              <ion-label>
                <h3>
                   test 
                </h3>
                <span
                  class="claim-number"
                >
                  Claim #12345
                </span>
                <br />
                <span
                  class="status"
                >
                  Status: active
                </span>
                <br />
                <span
                  class="subtitle"
                >
                   Lamborghini Gallardo 
                </span>
              </ion-label>
              <csaa-theme-icon
                slot="end"
              >
                <img
                  src="/assets/csaa-mobile/vectors/icn-Auto.svg"
                  style="width: 40px; height: 40px;"
                />
              </csaa-theme-icon>
            </ion-item>
          </ion-list>
          <ion-button
            expand="full"
          >
            I Had an Accident
          </ion-button>
        </ion-card-content>
      </ion-card>
    </csaa-claims-card>
  </div>
</test-host-component>
`;
